timeseries/tools/ingestion/perfomance_testing_202509.md

```diff
@@ -0,0 +1,255 @@
+    return analysed_df
+
+def summarise_df(df):
+    summary_cols = [
+        "total_time_per_message_ms",
+        "total_time_ms",
+        "ingress_time_ms",
+        "dedup_time_ms",
+        "validation_time_ms",
+        "writing_time_ms",
+    ]
+    summary_df = df[summary_cols].describe()
+    return summary_df
+
+
+def calculate_projected_times(df):
+    total_mean_time_per_message = df.filter(df["statistic"] == "mean")[
+        "total_time_per_message_ms"
+    ][0]
+
+    data = [
+        {
+            "label": "1 message",
+            "num_messages": 1,
+            "total_time": total_mean_time_per_message,
+        },
+        {
+            "label": "1 day",
+            "num_messages": 1440,
+            "total_time": total_mean_time_per_message * 1440,
+        },
+        {
+            "label": "1 week",
+            "num_messages": 10080,
+            "total_time": total_mean_time_per_message * 10080,
+        },
+        {
+            "label": "1 month (31 days)",
+            "num_messages": 44640,
+            "total_time": total_mean_time_per_message * 44640,
+        },
+    ]
+
+    projected_df = pl.DataFrame(data)
+    projected_df = projected_df.with_columns(
+        pl.from_epoch("total_time", time_unit="ms").dt.time().alias("total_time")
+    )
+    return projected_df
+
+
+if __name__ == "__main__":
+    main()
+
+```
+</details>
+
+## Results
+
+### Batches containing single messages only
+For a single message per batch, over 546 batches, the mean time taken per batch was 323ms. Extrapolating this out to
+estimate the time required for 1 day, 1 week and 1 month (31 days) worth of messages for a single site produces the 
+following numbers:
+
+|           | Num messages | total_time   |
+|-----------|--------------|--------------|
+| 1 message	| 1            | 00:00:00.323 |
+| 1 day     | 1440         | 00:07:45.684 |
+| 1 week    | 10080        | 00:54:19.790 |
+| 1 month   | 44640        | 04:00:36.216 |
+
+When looking at the proportion of time taken by each stage within a batch, it is clear that writing is the most time consuming (~57%), followed by the initial ingress stage (~37%). Validation (shown in yellow in the chart below) is negligible (~0.4%), and whilst deduplication takes longer than validation, it also only contributes a minor amount to the total processing time per batch (~6%).
+
+<img src="./images/performance_testing/mean_proportions_within_a_batch.png" alt="Mean proportion of time taken per stage within a batch" height="400px">
+
+### Batches containing multiple messages
+
+Looking at the same set of log messages, but without excluding multiple-message batches allows any economies of scale to become visible. The chart below shows both the total time per batch and the calculated total time per message within a batch, and clearly shows that whilst the total time per batch increases with more messages, on a per-message basis, each message is processed faster within the batch. 
```

Comment:
A small thing - the text talks about increasing the number of messages per batch, but the title on the graph talks about number of sites.  I think I understood it once I got to the final paragraph of your document, where it essentially says that the number of messages per batch = number of sites per batch.  Not sure if it needs clarifying or if its my mis-understanding